### PR TITLE
[DO-NOT-MERGE] Changes for Revision

### PR DESCRIPTION
**Pull request by acamino**
**Don't merge this branch**. This pull request exists only for code review purposes.

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -0,0 +1,4 @@
+export TWILIO_ACCOUNT_SID=my-sid
+export TWILIO_AUTH_TOKEN=my-token
+export TWILIO_PHONE_NUMBER=my-phone-number
+export APP_SECRET=my-secret
```
Comment (acamino) [inline .env.example:4]: 👍

File: app.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+app.use('/enable-tfa-via-app/', require('./routes/enable-tfa-via-app-route'));
+app.use('/enable-tfa-via-sms/', require('./routes/enable-tfa-via-sms-route'));
+app.use('/logout/', require('./routes/logout-route'));
+app.use('/sign-up/', require('./routes/sign-up-route'));
+app.use('/user/', require('./routes/user-route'));
+app.use('/verify-tfa/', require('./routes/verify-tfa-route'));
+app.use('/auth-qr-code.png/', require('./routes/auth-qr-code-png-route'));
+
+// catch 404 and forward to error handler
+app.use((req, res, next) => {
```
Comment (acamino) [inline app.js:39]: +1 for using arrow functions.

File: blog-post.txt [hunk 3]
```diff
@@ -0,0 +1,684 @@
...
+  <li><b>Adding Two-Factor Authentication</b>
+Now that we know more about how TOTP works, I'll show you how to add it to the example application that we started with.</li>
+  <li><b>Adding Google Authenticator</b>
+Here I will show you how to enable your users to authenticate via the Google Authenticator.</li>
+  <li><b>Adding Twilio</b>
+Here I will show you how to enable your users to authenticate using a code that is delivered to their phone via SMS.</li>
+  <li><b>Try It Out Yourself</b>
+Finally, I give you a working example of a site that can use both the Google Authenticator and SMS to do two-factor authentication.</li>
+</ol>
+&nbsp;
```
Comment (acamino) [inline blog-post.txt:30]: Not sure if this `&nbsp;` is needed, could you double check please?

File: blog-post.txt [hunk 4]
```diff
@@ -0,0 +1,684 @@
...
+
+A great pattern that we are seeing for implementing <a href="http://www.twilio.com/solutions/two-factor-authentication">two-factor authentication</a> is to use the TOTP (Time-based One-time Password Algorithm) standard for the second authentication step. What is so cool about TOTP is that it is flexible enough to allow your users to generate their authentication tokens directly on their smart phones using a TOTP app like <a href="http://support.google.com/accounts/bin/answer.py?hl=en&amp;answer=1066447">Google Authenticator</a> or have their tokens sent to their mobile phone via SMS.
+
+This means that you only need to implement and test one additional authentication scheme, but get the benefits of having two different ways that your users can get tokens.
+
+The best way to see how this is done is to look at some code. Let's do that now.
+
+Here are the topics that I'll be covering:
+<ol>
```
Comment (acamino) [inline blog-post.txt:16]: +1 for fixing the ordered list.

File: models/user.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
+'use strict';
+
+var bcrypt = require('bcrypt-nodejs')
+  , totp = require('../lib/totp')
+  , mongoose = require('mongoose')
+  , base32 = require('thirty-two');
+
+var schema = new mongoose.Schema({
+  username: { type: String, unique: true, required: true, dropDups: true },
+  totp_secret: String,
```
Comment (acamino) [inline models/user.js:10]: Is there any reason why we're using _snake_case_?

Please take a look at:
- http://mongoosejs.com/docs/schematypes.html
- http://mongoosejs.com/docs/models.html

Seems like the preferred style is _camelCase_. What do you think?

File: package.json [hunk 6]
```diff
@@ -0,0 +1,38 @@
+{
+  "name": "blog-tfa-node",
+  "version": "0.0.0",
+  "private": true,
+  "scripts": {
+    "start": "node ./bin/www",
+    "test": "export NODE_ENV=test && export TWILIO_ACCOUNT_SID=my-account-sid && export TWILIO_AUTH_TOKEN=my-auth-token && export TWILIO_NUMBER=my-twilio-number && export APP_SECRET=your-secret && mocha spec --recursive",
```
Comment (acamino) [inline package.json:7]: Nitpick: To do not pollute your environment with env variables, you can change this line by:

``` javascript
"test": "NODE_ENV=test TWILIO_ACCOUNT_SID=my-account-sid TWILIO_AUTH_TOKEN=my-auth-token TWILIO_NUMBER=my-twilio-number APP_SECRET=your-secret mocha spec --recursive"
```

File: public/css/bootstrap-responsive.css [hunk 7]
```diff
@@ -0,0 +1,1109 @@
+/*!
```
Comment (acamino) [inline public/css/bootstrap-responsive.css:1]: Not sure about copying _bootstrap-responsive.css_ in the project. What about using a CDN to resolve this dependency?

Comment (acamino) [inline public/css/bootstrap-responsive.css:1]: Ditto for all _bootstrap_ dependencies.

File: routes/auth-qr-code-png-route.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
+'use strict';
```
Comment (acamino) [inline routes/auth-qr-code-png-route.js:1]: This file is already in the _routes_ folder, do you think is necessary to add the _route_ suffix to the file name. The same criteria should applied to all the routes.

File: routes/enable-tfa-via-sms-route.js [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+  , loginRequired = require('../lib/login-required')
+  , buildData = require('../lib/build-data');
+
+// GET /enable-tfa-via-sms/
+router.get('/', loginRequired, (req, res, next) => {
+  var data = buildData(req);
+  res.render('enable_tfa_via_sms.pug', data);  
+});
+
+// POST //enable-tfa-via-sms/
```
Comment (acamino) [inline routes/enable-tfa-via-sms-route.js:16]: Love the comments, but this one looks slightly different.

File: routes/main-page-route.js [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    var data = buildData(req);
+    if (!user) {
+      data.opts.invalidUsernameOrPassword = true;
+      res.render('main_page.pug', data);
+    } else {
+      user.validatePassword(req.body.password, (err, isValid) => {
+        if (!isValid) {
+          data.opts.invalidUsernameOrPassword = true;
+          res.render('main_page.pug', data);
+        } else {
```
Comment (acamino) [inline routes/main-page-route.js:27]: What about using `if else` here?

File: spec/app.spec.js [hunk 11]
```diff
@@ -0,0 +1,568 @@
...
+  describe('GET /', () => {
+    it('responds with 200 OK', (done) => {
+      testSession.get('/')
+      .expect((res) => {
+          var $ = cheerio.load(res.text);
+          expect($('.well div h1').text()).to.equal('Don\'t have an account?');
+          expect($('button.btn').text()).to.equal('Log in');
+          expect($('a.btn').text()).to.contain('Sign up');
+          var text = 'This is a demonstration of how to add TOTP based ' +
+                      'Two-Factor Authentication to an existing application.';
```
Comment (acamino) [inline spec/app.spec.js:40]: Kinda picky but could you please align this line?

File: spec/app.spec.js [hunk 12]
```diff
@@ -0,0 +1,568 @@
...
+        })
+      .expect(200, done);
+    });
+  });
+
+});
+
+describe('sign in', () => {
+    
+  describe('with badpassword', () => {
```
Comment (acamino) [inline spec/app.spec.js:51]: Typo.

File: .env.example [hunk 13]
```diff
@@ -0,0 +1,4 @@
+export TWILIO_ACCOUNT_SID=my-sid
```
Comment (acamino) [inline .env.example:1]: I strongly recommend you to use a https://www.npmjs.com/package/dotenv. 
_.env.example_ will change slightly.

File: README.md [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+   $ git clone git@github.com:TwilioDevEd/blog-tfa-node.git
+   $ cd blog-tfa-node
+   ```
+
+1. Copy the sample configuration file and edit it to match your configuration.
+
+  ```bash
+  $ cp .env.example .env
+  ```
+
```
Comment (acamino) [inline README.md:33]: If you use https://www.npmjs.com/package/dotenv there is no need to add an additional step like `source .env`.

File: views/layout.pug [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+    link(rel='apple-touch-icon-precomposed', sizes='72x72', href='/static/ico/apple-touch-icon-72-precomposed.png')
+    link(rel='apple-touch-icon-precomposed', href='/static/ico/apple-touch-icon-57-precomposed.png')
+  body
+    .navbar.navbar-fixed-top.navbar-inverse
+      .navbar-inner
+        .container
+          a.brand(href='/')
+            | Two-Factor Authentication Example
+          ul.nav
+            if opts.isAuthenticated
```
Comment (acamino) [inline views/layout.pug:24]: What would happen if `opts` is undefined?

File: package.json (hunk 6 above)
Comment (mosampaio) [inline package.json:7]: Nice! I did't know about it.

File: models/user.js (hunk 5 above)
Comment (mosampaio) [inline models/user.js:10]: +1

File: public/css/bootstrap-responsive.css (hunk 7 above)
Comment (mosampaio) [inline public/css/bootstrap-responsive.css:1]: +1

File: routes/main-page-route.js (hunk 10 above)
Comment (mosampaio) [inline routes/main-page-route.js:27]: did you mean _else if_?

File: spec/app.spec.js (hunk 11 above)
Comment (mosampaio) [inline spec/app.spec.js:40]: ok

File: routes/main-page-route.js (hunk 10 above)
Comment (acamino) [inline routes/main-page-route.js:27]: You right. :sweat_smile:

File: views/layout.pug (hunk 15 above)
Comment (mosampaio) [inline views/layout.pug:24]: It breaks, I will fix that

File: spec/app.spec.js (hunk 12 above)
Comment (mosampaio) [inline spec/app.spec.js:51]: ok

File: .env.example (hunk 13 above)
Comment (mosampaio) [inline .env.example:1]: gotcha

File: README.md (hunk 14 above)
Comment (mosampaio) [inline README.md:33]: ok

File: routes/enable-tfa-via-sms-route.js (hunk 9 above)
Comment (mosampaio) [inline routes/enable-tfa-via-sms-route.js:16]: good catch

**Comment by acamino** [comment]
Thanks @mosampaio!

**Pull request closed without merging**